### PR TITLE
[docs] Editing pass over Dataset docs

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This PR makes a general editing pass over the Dataset docs. In particular:
- Move scheduling with Tune section out of key concepts and into a user guide
- Combine scheduling / memory management into a combined guide
- Coalesce pipeline usage into one guide and move advanced examples into example section
- Some rearrangements to the order and titles of the sections for more natural flow
- Editing to remove out-dated / unnecessary details from various sections

Depends on: https://github.com/ray-project/ray/pull/26934

### CODE REVIEW

File: doc/source/data/performance-tips.rst [hunk 1]
```diff
@@ -7,14 +7,7 @@ Debugging Statistics
...
 You can view debug stats for your Dataset and DatasetPipeline executions via ``ds.stats()``.
-These stats can be used to understand the performance of your Datasets workload and can help you debug problematic bottlenecks.
-
-At a high level, execution stats for tasks (e.g., CPU time) are attached to block metadata objects.
-Datasets have stats objects that hold references to these stats and parent dataset stats (this avoids stats holding references to parent datasets, allowing them to be garbage collected).
-Similarly, DatasetPipelines hold stats from recently computed datasets.
-In addition, we also collect statistics about iterator timings (time spent waiting / processing / in user code).
-Here's a sample output of getting stats in one of the most advanced use cases,
-namely iterating over a split of a dataset pipeline in a remote task:
+These stats can be used to understand the performance of your Dataset workload and can help you debug problematic bottlenecks. Note that both execution and iterator statistics are available:
```
Comment (jianoaix) [inline doc/source/data/performance-tips.rst:10]: Can we briefly talk about how to read the stats?
For example, "Remote wall time": is it the time to complete all tasks in a stage, or just one of them? "Remote CPU time": similarly, is it the sum of all CPU costs across tasks or just a single one?

**Review by richardliaw — APPROVED** [review]
stamping

File: doc/source/data/dataset-internals.rst [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+returns a Dataset whose all subsequent operations will be lazy. These operations
+won't be executed until the dataset is consumed or
+:meth:`ds.fully_executed() <ray.data.Dataset.fully_executed>` is called to manually
+trigger execution.
+
+Stage Fusion Optimization
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to reduce memory usage and task overheads, Datasets will automatically fuse together
+lazy operations that are compatible:
```
[the remark this answers is not in the document]
Comment (ericl) [reply doc/source/data/dataset-internals.rst:85]: This is covered already by compute (you can't specify actors for a non-class UDF), right?

File: doc/source/data/pipelining-compute.rst [hunk 3]
```diff
@@ -202,3 +201,88 @@ You can also specify the size of each window using ``bytes_per_window``. In this
         .read_binary_files("s3://bucket/image-dir") \
         .window(bytes_per_window=10e9)
     # -> INFO -- Created DatasetPipeline with 73 windows: 9120MiB min, 9431MiB max, 9287MiB mean
+    # -> INFO -- Blocks per window: 10 min, 16 max, 14 mean
+    # -> INFO -- ✔️  This pipeline's per-window parallelism is high enough to fully utilize the cluster.
+    # -> INFO -- ✔️  This pipeline's windows can each fit in object store memory without spilling.
```
[the remark this answers is not in the document]
Comment (ericl) [reply doc/source/data/pipelining-compute.rst:206]: Yeah, though the common case is a 1-2 stages due to stage fusion. We could detect multi-stage cases in the future, but so far they seem pretty rare.

File: doc/source/data/pipelining-compute.rst [hunk 4]
```diff
@@ -202,3 +201,88 @@ You can also specify the size of each window using ``bytes_per_window``. In this
...
+**Pipeline**:
+
+.. image:: images/dataset-repeat-2.svg
+
+Handling Epochs
+~~~~~~~~~~~~~~~
+
+It's common in ML training to want to divide data ingest into epochs, or repetitions over the original source dataset.
+DatasetPipeline provides a convenient ``.iter_epochs()`` method that can be used to split up the pipeline into epoch-delimited pipeline segments.
+Epochs are defined by the last call to ``.repeat()`` in a pipeline, for example:
```
[the remark this answers is not in the document]
Comment (ericl) [reply doc/source/data/pipelining-compute.rst:257]: If a dataset is repeated twice, the "source dataset" is defined by the second repeat, not the first.

File: doc/source/data/performance-tips.rst (hunk 1 above)
Comment (ericl) [reply doc/source/data/performance-tips.rst:10]: Sure, I think that would be a good followup.

**Review by scv119 — APPROVED** [review]

**Comment by ericl** [comment]
I'm going to try to merge this for now. We still need a lot of improvements on Datasets, but this will lay the groundwork for further fine-tuning that we can parallelize.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**